src/python/pants/jvm/target_types.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    Dependencies,
+    Sources,
+    StringSequenceField,
+    Target,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class JavaSources(Sources):
+    expected_file_extensions = (".java",)
+
+
+class JavaTestsSources(JavaSources):
+    default = ("*Test.java",)
+
+
+class JunitTests(Target):
+    alias = "junit_tests"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        Dependencies,
+        JavaTestsSources,
+    )
+    help = "Java tests, run with Junit."
+
+
+class JavaLibrarySources(JavaSources):
+    default = ("*.java",) + tuple(f"!{pat}" for pat in JavaTestsSources.default)
+
+
+class JavaLibrary(Target):
+    alias = "java_library"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        Dependencies,
+        JavaLibrarySources,
+    )
+    help = "Java source code."
+
+
+class MavenRequirementsField(StringSequenceField):
+    alias = "maven_requirements"
+    help = (
+        "A sequence of Maven coordinate strings, e.g. ['org.scala-lang:scala-compiler:2.12.0', "
+        "'org.scala-lang:scala-library:2.12.0']."
+    )
+
+
+class JvmLockfileSources(Sources):
+    expected_file_extensions = (".lockfile",)
+    expected_num_files = range(
+        2
+    )  # NOTE: This actually means 0 or 1 files; `range`'s end is noninclusive.
+    help = "A single Pants Coursier Lockfile source.  Use `./pants coursier-resolve ...` to generate (or regenerate) the Lockfile.  If the Lockfile doesn't exist on disk, the first run of `coursier-resolve` will attempt to generate it for you to the default file name ('coursier_resolve.lockfile').  After running `coursier-resolve` for the first time, you should update this field's `sources` to explicit take ownership of the generated lockfile."
```

Comment:
Preferred to split this over multiple lines using soft wrapping like you do on lines 53-56. Note that Black will not format strings out of the risk of breaking them. While it's not a hard requirement, we do try to still stick with the 100 char line limit.

Reply:
Also, probably helpful to add `\n\n` between the first line and the rest of the sentences so that you have two paragraphs. We've been doing that more recently to make things easier to read.

Reply:
Thanks for the great help message!